### PR TITLE
Feature: Retry orderbook when watchMarket fails

### PR DESCRIPTION
**Pull request by treygriffith**
## Description
This PR allows the Broker to stay up when `watchMarket` fails, and retry `watchMarket` until it comes back up. It throws an error if trying to use the orderbook which relies on being synced with the Relayer.

## Todos
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: broker-daemon/orderbook/index.js [hunk 1]
```diff
@@ -130,13 +144,56 @@ class Orderbook {
...
+      MarketEvent.rangeFromTimestamp(timestamp)
+    )
+  }
+
+  /**
+   * Ensures that the orderbook is synced before accessing it
+   * @return {void}
+   * @throws {Error} If Orderbook is not synced to Relayer
+   */
+  _assertSynced () {
```
Comment (dannypaz) [inline broker-daemon/orderbook/index.js]: [nit] Is this a convention we want to use going forward? I remember having a convo about this a few months ago and we decided to omit `underscores`. If we wanted a truly private method, we could add the function to the top of this file, and not export it.

File: broker-daemon/relayer/market-watcher.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+class MarketWatcher extends EventEmitter {
+  /**
+   * Set up the watcher
+   * @param  {grpc.ServerStreaming} watcher        Stream of events from the Relayer
+   * @param  {Sublevel}             store          Leveldb compatible store
+   * @param  {Object}               RESPONSE_TYPES Response types from the proto file
+   * @param  {Object} logger
+   * @return {MarketWatcher}
+   */
+  constructor (watcher, store, RESPONSE_TYPES, logger) {
```
Comment (dannypaz) [inline broker-daemon/relayer/market-watcher.js]: should we just call this `stream` instead of watcher?

File: broker-daemon/orderbook/index.js [hunk 3]
```diff
@@ -28,19 +28,30 @@ class Orderbook {
...
+      this.bidIndex = await (new BidIndex(this.store)).ensureIndex()
 
-    return this.logger.info(`Market ${this.marketName} initialized.`)
+      this.synced = true
+      this.logger.info(`Market ${this.marketName} initialized.`)
+    })
+
+    watcher.on('end', (error) => {
+      this.logger.info(`Market ${this.marketName} unavailable, retrying in 5s`, { error })
+      setTimeout(() => {
```
Comment (dannypaz) [inline broker-daemon/orderbook/index.js]: TODO for the exponential backoff stuff from @martineehrlich

File: broker-daemon/orderbook/index.js [hunk 4]
```diff
@@ -150,7 +211,7 @@ class Orderbook {
 
   /**
    * Gets the last time this market was updated with data from the relayer
-   *
+   * @private
```
Comment (martineehrlich) [inline broker-daemon/orderbook/index.js]: lol

File: broker-daemon/relayer/market-watcher.js [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    super()
+    this.watcher = watcher
+    this.store = store
+    this.migrating = null
+    this.logger = logger
+    this.RESPONSE_TYPES = RESPONSE_TYPES
```
Comment (martineehrlich) [inline broker-daemon/relayer/market-watcher.js]: I am confused as to why you are passing this in, when we have hard coded expected response types lower in the file?

Comment (treygriffith) [reply broker-daemon/relayer/market-watcher.js]: They aren't hardcoded lower in the file, we're pulling them out of the proto file in relayer-client.js

Comment (dannypaz) [reply broker-daemon/relayer/market-watcher.js]: @treygriffith when we upgrade the grpc on broker, we will no longer have access to those enums from the protofile. As a preventative measure, I would hardcode them somewhere

File: broker-daemon/orderbook/index.js [hunk 6]
```diff
@@ -8,6 +8,8 @@ const nano = require('nano-seconds')
 const consoleLogger = console
 consoleLogger.debug = console.log.bind(console)
 
+const RETRY_WATCHMARKET = 5000
```
Comment (dannypaz) [inline broker-daemon/orderbook/index.js]: Add docs

File: broker-daemon/relayer/market-watcher.js (hunk 2 above)
Comment (dannypaz) [reply broker-daemon/relayer/market-watcher.js]: @treygriffith Wondering if there is a more descriptive term for this object `watcher`

File: broker-daemon/relayer/market-watcher.js (hunk 5 above)
Comment (dannypaz) [reply broker-daemon/relayer/market-watcher.js]: NVM, we can hold off until the actual migration. per our talk @treygriffith

**Pull request merged**